File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    for p in permutations(dungeons, len(dungeons)):
        cnt = 0
        ph = k
        for dungeon in p:
            if ph >= dungeon[0]:
                cnt += 1
                ph -= dungeon[1]
        if answer < cnt:
            answer = cnt
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    static boolean[] visit;
    static int answer = 0;
    public int solution(int k, int[][] dungeons) {
        visit = new boolean[dungeons.length];        
        dfs(k, 0, dungeons);
        return answer;
    }
    
    static void dfs(int tired, int cnt, int[][] dungeons) {
        for (int i = 0; i < dungeons.length; i++) {
            if (!visit[i] && tired >= dungeons[i][0]) {
                visit[i] = true;
                dfs(tired-dungeons[i][1], cnt+1, dungeons);
                visit[i] = false;
            }
        }
        
        answer = Math.max(answer, cnt);
    }
}